flutter: Construct SDK from Git repository + `flutter precache`

hacker1024 opened the pull request:
## Description of changes

This PR rethinks the approach to building Flutter and packaging artifacts.

### Existing issues

1. A different Flutter SDK tarball is required for every platform, even though the differences between them are minimal. This is a waste of resources, especially for maintainers and the official binary cache.
2. SDK tarballs contain components we do not need, such as the Dart SDK and pre-cached Dart packages. This is a significant waste of space.
3. Keeping track of all the engine artifact behaviour is challenging. Our current system needs to re-implement much of the logic from the Flutter tool in Nix, often with [ugly hacks](https://github.com/NixOS/nixpkgs/blob/0a617df61a72f4883ecc4bea2fdb6a9b92d62565/pkgs/development/compilers/flutter/engine-artifacts/default.nix#L158) as it changes - Dart can be refactored a lot more easily than Nix can in this case.
4. We have not yet packaged non-engine artifacts, which have even more flexible installation processes. If we ever want more control over them, working them in to the existing system will be very hard.

### Changes in this PR

- The Flutter SDK is set up directly from the Git repository - no tarballs required.
- The Flutter CLI is built in a standard way, with `buildDartApplication`.
- All artifacts are fetched with `flutter precache` and packed into fixed-output derivations (one per platform).

With these changes, we save a lot of resources, reduce the build complexity, and remove the need to keep up with the artifact logic in the Flutter tool.

### Drawbacks

- There is less granularity over artifact inclusion. In practice, this shouldn't change much - you need most of the platform artifacts to develop for a platform anyway.
- The Flutter SDK no longer knows its own version, as there is no way to reproducibly keep `.git` directory. This leads to output from `flutter doctor` as shown below.

```
[!] Flutter (Channel [user-branch], 0.0.0-unknown, on NixOS 23.11 (Tapir) 6.5.7-zen1, locale en_AU.UTF-8)
    ! Flutter version 0.0.0-unknown on channel [user-branch] at /nix/store/h7bbwwy1hmmv2l2d7y1pf6h0165l04z4-flutter-wrapped-3.13.8-sdk-links
      Currently on an unknown channel. Run `flutter channel` to switch to an official channel.
      If that doesn't fix the issue, reinstall Flutter by following instructions at https://flutter.dev/docs/get-started/install.
      Cannot resolve current version, possibly due to local changes.
      Reinstall Flutter by following instructions at https://flutter.dev/docs/get-started/install.
    ! Unknown upstream repository.
      Reinstall Flutter by following instructions at https://flutter.dev/docs/get-started/install.
    • Framework revision ab75bef3f9 (54 years ago), 1970-01-01 00:00:00 +0000
    • Engine revision 767d8c75e8
    • Dart version 3.1.4
    • DevTools version 2.25.0
    • If those were intentional, you can disregard the above warnings; however it is recommended to use "git" directly to perform update checks and upgrades.
```

### TODO

- ~~Web support is broken for now. WIP.~~

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

@NixOS/flutter

hacker1024 committed:
flutter: Build flutter_tools with buildDartApplication

hacker1024 committed:
flutter: Use deepClone = true for SDK repository

hacker1024 committed:
flutter: Include patches in flutter_tools snapshot

hacker1024 committed:
flutter: Remove hard dependency on Git

hacker1024 committed:
flutter: Allow overriding the FLUTTER_CACHE_DIR

hacker1024 committed:
flutter: Use custom launchers

hacker1024 committed:
flutter: Specify commit date for fake SDK repository

hacker1024 committed:
flutter: Remove any existing Dart SDK

hacker1024 committed:
flutter: Use JIT snapshot for flutter_tools

hacker1024 committed:
flutter: Remove JIT explanation

Turns out it was a patch that was causing the strange behaviour.

hacker1024 committed:
flutter: Disable automatic updates without breaking precache

hacker1024 committed:
flutter: Allow adding additional Dart VM options when launching CLI

hacker1024 committed:
flutter: Deregister the Pub dependencies artifact

There is no need to cache the flutter_tool pub dependencies, as it is built reproducibly with Nix.

hacker1024 committed:
flutter: Update FLUTTER_ROOT in SDK symlink derivation

hacker1024 committed:
flutter: Fetch artifacts with `flutter prefetch`

hacker1024 committed:
flutter: Propagate .git in SDK symlink derivation

hacker1024 committed:
flutter: Pass pubGetScript through to buildDartApplication

hacker1024 committed:
fetchDartDeps: Use given setup and Pub get script in dependency list derivation

hacker1024 committed:
buildDartApplication: Add entire Dart derivation to nativeBuildInputs

mkg20001 commented:
can the flutter build code be patched to get the version from an environment variable given at build time?

hacker1024 commented:
> can the flutter build code be patched to get the version from an environment variable given at build time?

I looked into this, but the Flutter version code is actually quite complicated and involves large classes with numerous fields. A patch would be quite intrusive, and doesn't seem worth maintaining.

hacker1024 committed:
fetchDartDeps: Generalise dependency derivation logic

hacker1024 committed:
fetchDartDeps: Pass through the dependency cache

hacker1024 committed:
fetchDartDeps: Add a package_config.json derivation

hacker1024 committed:
flutter: Link in the fluter_tools package_config.json

hacker1024 commented:
I've fixed the Web support. Turns out that some of `flutter_tools`'s dependencies include static assets that the application tries to find by reading its own `package_config.json`, so that has to be made available along with the Pub package cache used during compilation.

hacker1024 committed:
flutter: Add platform artifact derivations to cacheDir passthru

hacker1024 committed:
flutter: Set host platform at compilation

The Flutter CLI normally detects the host platform at runtime, but this results in incorrect behaviour in the Nix build environment.

FlafyDev reviewed (commented):
Nice! That's a way cleaner way of fetching the artifacts. didn't know flutter had that command.
I left a few comments, but I might add more later since this pr is quite big and I still haven't had the time to see all of it

Also, thanks to this PR, we can remove the FLUTTER_CACHE_DIR variable.

1. remove its section from move-cache.patch
2. remove the line to set it in the immutableFlutter wrapper
3. in `fetch-artifacts.nix`, do this instead:
```
  lndir -silent ${flutter} .
  HOME="$NIX_BUILD_TOP" FLUTTER_ROOT="." flutter precache -v '--${platform}' ${builtins.concatStringsSep " " (map (p: "'--no-${p}'") (lib.remove platform platforms))}
  find ./bin/cache -type l -exec rm {} \;
  cp -r ./bin/cache $out
```

hacker1024 commented on pkgs/development/compilers/flutter/flutter.nix:
```diff
@@ -34,55 +44,41 @@ let
+        # The flutter_tools package tries to run many Git commands. In most
+        # cases, unexpected output is handled gracefully, but commands are never
+        # expected to fail completely. A blank repository needs to be created.
+        if [ ! -d .git ]; then
```
It can happen if one uses a local checkout of the Flutter as the source without filtering the .git directory.

FlafyDev replied:
Then should we do something in that case? Or do we leave it up to the new source to make its .git reproducible?
I think it should remove any .git directory so that the derivation will always make its own.

hacker1024 replied:
I have changed it to remove existing `.git` directories.

hacker1024 commented on pkgs/development/compilers/flutter/artifacts/prepare-artifacts.nix:
`fetch-artifacts.nix` is responsible for generating a fixed-output derivation. This should do as little as possible, to keep the hashes stable. The output contents cannot reference any Nix store paths.

`prepare-artifacts` patches the prebuilt artifacts so that they are actually usable. It changes their directory structure, and it extended with more complex steps - for example, the Linux artifact derivation adds `gtk3` as a `buildInput` for `patchelf`.

FlafyDev replied:
I see, that makes sense

hacker1024 commented on pkgs/development/compilers/flutter/wrapper.nix:
```diff
@@ -120,15 +79,15 @@ let
   # We do not patch it since the script doesn't require engine artifacts(which are the only thing not added by the unwrapped derivation), so it shouldn't fail, and patching it will just be harder to maintain.
   immutableFlutter = writeShellScript "flutter_immutable" ''
```
As far as I know, the C implementation of `makeWrapper` (which is being used currently) does not support using environment variables such as `HOME`.

FlafyDev replied:
Oh I didn't know that. Also, where in the code do we specify to use the C implementation? is it by default now?

hacker1024 replied:
I'm not sure. It may have something to do with `wrapGAppsHook`.

hacker1024 commented:
> Nice! That's a way cleaner way of fetching the artifacts. didn't know flutter had that command. I left a few comments, but I might add more later since this pr is quite big and I still haven't had the time to see all of it
> 
> Also, thanks to this PR, we can remove the FLUTTER_CACHE_DIR variable.
> 
>     1. remove its section from move-cache.patch
> 
>     2. remove the line to set it in the immutableFlutter wrapper
> 
>     3. in `fetch-artifacts.nix`, do this instead:
> 
> 
> ```
>   lndir -silent ${flutter} .
>   HOME="$NIX_BUILD_TOP" FLUTTER_ROOT="." flutter precache -v '--${platform}' ${builtins.concatStringsSep " " (map (p: "'--no-${p}'") (lib.remove platform platforms))}
>   find ./bin/cache -type l -exec rm {} \;
>   cp -r ./bin/cache $out
> ```

Nice, I was looking for a way to do this without accidentally including SDK files in the FOD. 

My only concern with this method is that it may break if `precache` makes any links of its own, which it could technically do. This can be improved to only remove links going to the base `flutter` derivation, though.

hacker1024 committed:
flutter: Add artifact hashes for all platforms

hacker1024 committed:
flutter: Move artifact override logic to fetch-artifacts.nix

hacker1024 commented on pkgs/development/compilers/flutter/wrapper.nix:
```diff
@@ -57,61 +44,33 @@
+    }).overrideAttrs (
+      if builtins.pathExists ./artifacts/overrides/${platform}.nix
+      then callPackage ./artifacts/overrides/${platform}.nix { }
+      else ({ ... }: { })
```
I have moved the overriding logic over. The hash selection remains in the wrapper, to allow people to supply their own hashes by overriding it.

hacker1024 committed:
flutter: Remove existing .git directories

hacker1024 committed:
flutter: Remove runHooks in fetch-artifacts.nix runCommand

hacker1024 committed:
flutter: Remove FLUTTER_CACHE_DIR

hacker1024 commented:
I have implemented the proposed idea to remove `FLUTTER_CACHE_DIR`.

mkg20001 commented:
re version: another solution to fix the version wouldbe to create a mock git binary that just echos a preset output based on the arguments

or is the .git needed after build aswell?

hacker1024 commented:
> re version: another solution to fix the version wouldbe to create a mock git binary that just echos a preset output based on the arguments
> 
> or is the .git needed after build aswell?

It is needed at runtime, as the Flutter CLI uses it for version info all the time. We could add it to the wrapper, though.

That would be quite an effort though. A wide range of Git commands would need to be mocked.

FlafyDev commented on pkgs/development/compilers/flutter/patches/flutter3/move-cache.patch:
This patch is actually important. 
It is line 38 and after that we don't need anymore. (FLUTTER_CACHE_DIR stuff)

hacker1024 replied:
I don't think that it's necessary any more. AFAIK, the paths it patches are all looked up from `$FLUTTER_ROOT`.

I've added `--set FLUTTER_ALREADY_LOCKED true` to the launch wrapper to prevent writing to the immutable lockfile. With this, everything seems to work as expected.

FlafyDev commented on pkgs/development/compilers/flutter/artifacts/fetch-artifacts.nix:
```diff
@@ -41,6 +41,14 @@ runCommand "flutter-artifacts-${platform}"
+  mkdir "$FLUTTER_ROOT/bin/cache"
+
+  HOME="$(mktemp -d)" flutter precache -v '--${platform}' ${builtins.concatStringsSep " " (map (p: "'--no-${p}'") (lib.remove platform platforms))}
+  rm -r "$FLUTTER_ROOT/bin/cache/lockfile"
```
after `move-cache.patch`, this should be removed. (since the lockfile will be in ~/.cache)

hacker1024 committed:
flutter: Set FLUTTER_ALREADY_LOCKED to prevent writes to the immutable lockfile

hacker1024 committed:
flutter: Remove aarch64-darwin platform

There are no aarch64-darwin host artifacts.
Development should be done with Rosetta 2.

- https://github.com/flutter/flutter/issues/60118
- https://github.com/flutter/flutter/issues/69157
- https://github.com/flutter/flutter/issues/101138

hacker1024 commented:
I've removed support for `aarch64-darwin` for now, as there are no host artifacts for the platform. In theory, the Flutter tool could be built, but the artifacts will always have to be used through Rosetta 2.

It's likely that the artifact hashes in this case will match `x86_64-darwin`, in which case support can be re-introduced - but I have no hardware to test on.

hacker1024 committed:
fetchDartDeps: Use add milliseconds to date in package_config.json

This in compliance with the format specification.
https://github.com/dart-lang/language/blob/main/accepted/2.8/language-versioning/package-config-file-v2.md

FlafyDev commented on pkgs/development/compilers/flutter/artifacts/fetch-artifacts.nix:
```diff
@@ -0,0 +1,54 @@
+{ lib
+, runCommand
+, xorg
```
nit: can be replaced with `lndir` directly

FlafyDev commented on pkgs/development/compilers/flutter/patches/flutter3/disable-auto-update.patch:
Is flutter still using shared.sh? If not I think its section in this patch can be removed

FlafyDev reviewed (commented):
I think this PR is ready.
I left a few small comments.
Also, how do we get the hashes for the artifacts? It seems to depend on the host's platform to fetch them

maximoffua commented:
I have managed to get Flutter version working without `.git`: just fill out file `bin/cache/flutter.version.json`

```diff
diff --git a/pkgs/flutter/flutter.nix b/pkgs/flutter/flutter.nix
index 40aa090..69efe67 100644
--- a/pkgs/flutter/flutter.nix
+++ b/pkgs/flutter/flutter.nix
@@ -13,7 +13,6 @@
 , darwin
 , git
 , which
+, runCommand
 }:
 
 let
@@ -23,19 +22,6 @@ let
     inherit patches;
     inherit pubspecLockFile vendorHash depsListFile;
   };
+  formatTimestamp = format: (builtins.readFile (runCommand "timestamp" {} "echo -n `date '+${format}'` > $out"));
+  devToolsVersion = (builtins.fromJSON (builtins.readFile "${dart}/bin/resources/devtools/version.json")).version;
+  flutterVersion = {
+    inherit devToolsVersion;
+    flutterVersion = "${version}";
+    frameworkVersion = "${version}";
+    channel = "stable";
+    repositoryUrl = "https://github.com/flutter/flutter.git";
+    frameworkRevision = "nixpkgs000000000000000000000000000000000";
+    frameworkCommitDate = formatTimestamp "%Y-%m-%d %H:%M:%S";
+    engineRevision = "${engineVersion}";
+    dartSdkVersion = "${dart.version}";
+  };
 
   unwrapped =
     stdenv.mkDerivation {
@@ -65,10 +51,9 @@ let
         git init -b nixpkgs
         GIT_AUTHOR_NAME=Nixpkgs GIT_COMMITTER_NAME=Nixpkgs \
         GIT_AUTHOR_EMAIL= GIT_COMMITTER_EMAIL= \
+        GIT_AUTHOR_DATE='${formatTimestamp "%m/%d/%Y %H:%M:%S %z"}' \
+        GIT_COMMITTER_DATE='${formatTimestamp "%m/%d/%Y %H:%M:%S %z"}' \
-        GIT_AUTHOR_DATE='1/1/1970 00:00:00 +0000' GIT_COMMITTER_DATE='1/1/1970 00:00:00 +0000' \
           git commit --allow-empty -m "Initial commit"
         (. '${../../../build-support/fetchgit/deterministic-git}'; make_deterministic_repo .)
 
         mkdir -p bin/cache
 
@@ -84,7 +69,6 @@ let
         ln -s '${tools.dartDeps.packageConfig}' packages/flutter_tools/.dart_tool/package_config.json
 
         echo -n "${version}" > version
+        echo -n '${builtins.toJSON flutterVersion}' | tee bin/cache/flutter.version.json
       '';
 
       installPhase = ''
 ```

This is necessary, for any flutter project which specifies minimal SDK version.

maximoffua commented:
@hacker1024 could you please assist me with updating Flutter version in your `flutter-from-source` branch?

I have created Nix flake based on your work (see [here](https://github.com/maximoffua/flutter.nix)). But I cannot understand how to update Flutter version, specifically these hash sums of artifacts. Can you give me a hint of how to obtain a download URL and a proper hash for every component from `artifacts/hashes.nix` and files in `lockfiles/stable/`

FlafyDev commented:
> @hacker1024 could you please assist me with updating Flutter version in your `flutter-from-source` branch?
> 
> I have created Nix flake based on your work (see [here](https://github.com/maximoffua/flutter.nix)). But I cannot understand how to update Flutter version, specifically these hash sums of artifacts. Can you give me a hint of how to obtain a download URL and a proper hash for every component from `artifacts/hashes.nix` and files in `lockfiles/stable/`?

@maximoffua 
I'm not sure myself as well. But I did manage to make a nix script to print the hashes after some modifications to the PR.

```patch
diff --git a/pkgs/development/compilers/flutter/artifacts/fetch-artifacts.nix b/pkgs/development/compilers/flutter/artifacts/fetch-artifacts.nix
index d8b237cb1e16..9100f0939658 100644
--- a/pkgs/development/compilers/flutter/artifacts/fetch-artifacts.nix
+++ b/pkgs/development/compilers/flutter/artifacts/fetch-artifacts.nix
@@ -5,6 +5,7 @@
 , unzip
 
 , platform
+, archPlatform
 , flutter
 , hash
 }:
@@ -25,9 +26,13 @@ let
     # Use a version of Flutter with just enough capabilities to download
     # artifacts.
     supportedTargetPlatforms = [ ];
+
+    flutter = flutter.unwrapped.override {
+      inherit archPlatform;
+    };
   };
 in
-runCommand "flutter-artifacts-${platform}"
+runCommand "flutter-artifacts-${platform}-${archPlatform}"
 {
   nativeBuildInputs = [ xorg.lndir flutter' unzip ];
 
diff --git a/pkgs/development/compilers/flutter/flutter-tools.nix b/pkgs/development/compilers/flutter/flutter-tools.nix
index 4435c8fb6c76..cde029dba8a5 100644
--- a/pkgs/development/compilers/flutter/flutter-tools.nix
+++ b/pkgs/development/compilers/flutter/flutter-tools.nix
@@ -1,4 +1,4 @@
-{ hostPlatform
+{ archPlatform
 , buildDartApplication
 , git
 , which
@@ -35,7 +35,7 @@ buildDartApplication.override { inherit dart; } rec {
   '';
 
   dartEntryPoints."flutter_tools.snapshot" = "bin/flutter_tools.dart";
-  dartCompileFlags = [ "--define=NIX_FLUTTER_HOST_PLATFORM=${hostPlatform.system}" ];
+  dartCompileFlags = [ "--define=NIX_FLUTTER_HOST_PLATFORM=${archPlatform}" ];
 
   # The Dart wrapper launchers are useless for the Flutter tool - it is designed
   # to be launched from a snapshot by the SDK.
diff --git a/pkgs/development/compilers/flutter/flutter.nix b/pkgs/development/compilers/flutter/flutter.nix
index 1e68a3b02a9e..0d21b6a00839 100644
--- a/pkgs/development/compilers/flutter/flutter.nix
+++ b/pkgs/development/compilers/flutter/flutter.nix
@@ -11,6 +11,7 @@
 , darwin
 , git
 , which
+, archPlatform ? stdenv.hostPlatform.system
 }:
 
 let
@@ -19,6 +20,7 @@ let
     flutterSrc = src;
     inherit patches;
     inherit pubspecLock;
+    inherit archPlatform;
   };
 
   unwrapped =
diff --git a/pkgs/development/compilers/flutter/wrapper.nix b/pkgs/development/compilers/flutter/wrapper.nix
index 3dc67f4b6d7d..53aa489d74f6 100644
--- a/pkgs/development/compilers/flutter/wrapper.nix
+++ b/pkgs/development/compilers/flutter/wrapper.nix
@@ -51,6 +51,7 @@ let
       src = callPackage ./artifacts/fetch-artifacts.nix {
         inherit platform;
         flutter = callPackage ./wrapper.nix { inherit flutter; };
+        archPlatform = stdenv.hostPlatform.system;
         hash = artifactHashes.${platform}.${stdenv.hostPlatform.system} or "";
       };
     }));
```

this patch adds an `archPlatform` argument so this script is possible:

```nix
{pkgs ? import ./. {}}:
pkgs.callPackage (
  {
    callPackage,
    flutter,
    lib,
    symlinkJoin,
  }: let
    flutterPlatforms = [
      # "android"
      # "ios"
      # "web"
      "linux"
      # "windows"
      # "macos"
      # "fuchsia"
      # "universal"
    ];
    archPlatforms = [
      "x86_64-linux"
      "aarch64-linux"
      "x86_64-darwin"
      # "aarch64-darwin"
    ];

    # After packaging a new flutter version, put it here.
    newFlutter = flutter;

    derivations =
      lib.foldl' (
        acc: flutterPlatform:
          acc
          ++ (map (archPlatform:
            callPackage ./pkgs/development/compilers/flutter/artifacts/fetch-artifacts.nix {
              platform = flutterPlatform;
              inherit archPlatform;
              hash = "";
            })
          archPlatforms)
      ) []
      flutterPlatforms;
  in
    # Only way I found to bulid multiple derivations...
    symlinkJoin {
      name = "evaluate-derivations";
      paths = derivations;
    }
) {}
```

put it in a nix file at the root of nixpkgs and run 
```
nix build --keep-going --impure -L --expr 'import ./test.nix'
```
it should fail several times and print the hashes.

FlafyDev commented:
> I have managed to get Flutter version working without `.git`: just fill out file `bin/cache/flutter.version.json`
> 
> This is necessary, for any flutter project which specifies minimal SDK version.

@maximoffua Nice! I changed it a little so it doesn't cause [IFD](https://nixos.org/manual/nix/unstable/language/import-from-derivation).

And I don't think the use of `date` is really necessary..

```patch
diff --git a/pkgs/development/compilers/flutter/flutter.nix b/pkgs/development/compilers/flutter/flutter.nix
index 1e68a3b02a9e..c78589997b43 100644
--- a/pkgs/development/compilers/flutter/flutter.nix
+++ b/pkgs/development/compilers/flutter/flutter.nix
@@ -11,6 +11,7 @@
 , darwin
 , git
 , which
+, jq
 }:
 
 let
@@ -27,7 +28,7 @@ let
       inherit src patches version;
 
       buildInputs = [ git ];
-      nativeBuildInputs = [ makeWrapper ]
+      nativeBuildInputs = [ makeWrapper jq ]
         ++ lib.optionals stdenv.hostPlatform.isDarwin [ darwin.DarwinTools ];
 
       preConfigure = ''
@@ -67,6 +68,19 @@ let
         ln -s '${tools.pubcache}/package_config.json' packages/flutter_tools/.dart_tool/package_config.json
 
         echo -n "${version}" > version
+        cat <<EOF > bin/cache/flutter.version.json
+        {
+          "devToolsVersion": "$(cat "${dart}/bin/resources/devtools/version.json" | jq -r .version)",
+          "flutterVersion": "${version}",
+          "frameworkVersion": "${version}",
+          "channel": "stable",
+          "repositoryUrl": "https://github.com/flutter/flutter.git",
+          "frameworkRevision": "nixpkgs000000000000000000000000000000000",
+          "frameworkCommitDate": "1970-01-01 00:00:00",
+          "engineRevision": "${engineVersion}",
+          "dartSdkVersion": "${dart.version}"
+        }
+        EOF
       '';
 
       installPhase = ''
```

mkg20001 committed:
flutter: dont fail on missing lockfile

mkg20001 commented:
I'm currently preparing this for merge (rebase, etc). Anything missing or can I go ahead with merge?

FlafyDev commented:
> I'm currently preparing this for merge (rebase, etc). Anything missing or can I go ahead with merge?

I have a few things I want to add to the PR.

- version in flutter doctor/--version
- being able to update every hash for a potential update script

And a bit more stuff

Is there a way to add commits to this PR? Or should I make a new PR after this is merged?

mkg20001 commented:
I suppose we can do it after merge. If you are committer you can push directly to this branch in hacker1024's repo (maybe maintainer can push too, not sure). But I prefer that we add new stuff afterwards, as this PR is already quite big.
